feat: Skill point loadouts

ryanzhoudev opened the pull request:
~~(video coming soon)~~ 2 videos!

https://github.com/Wynntils/Artemis/assets/57310593/e019a501-b25f-46a8-9b19-7200e2c6ecd9

In text, video 1:
- Save loadouts (no armour/accessories) and builds (armour and/or accessories)
- Loadouts displayed with just name and points
- Builds displayed with name and armour, or accessories if no armour present
- Types can be converted between each other, description of how is in the convert button tooltip
- Confirmation when overwriting existing build/loadout
- Points can be very quickly, much faster than legacy. Loads in 5's where possible using shift-clicks.

https://github.com/Wynntils/Artemis/assets/57310593/a33bb9c8-62e4-41f9-9fad-bd786f516500

Video 2:
- All builds/loadouts display with number of total points and level requirement, in red when not met
- Load button has a warning message when level requirement not met
- Low level players without tome access are parsed properly

ryanzhoudev committed:
feat: Skill point loadouts

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
start making the model

ryanzhoudev committed:
try to fix accessory

ryanzhoudev committed:
add crafted note

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
fix sp parsing, spotless

ryanzhoudev committed:
close containers properly, tome parsing works

ryanzhoudev committed:
move total processing

ryanzhoudev committed:
crafted parsing

ryanzhoudev committed:
clean up loops

ryanzhoudev committed:
some debug stuff, fix total calculation race

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
Merge branch 'main' into sploadout

kristofbolyai committed:
refactor: In preparation for charm identifications, add support for custom stat type ranges, rework the inverted stat system, other fixes to internal rolls (#2282)

* refactor: Allow custom base ranges for items, rework inverted calculation, add option to display stat as inverted

* fix: Fix SpellStatType internal names for percentage values

* fix: Fix elemental defense stats

Turns out the API was wrong, and the removed type does not exist, and is only a typo. This is now confirmed to work in-game.

* fix: Correctly calculate internal role in every case

I've done throughout testing, but I think just running with the assertions enabled without any issues proves that this should work.

* feat: Reworked perfect/increase/decrease percentages

* fix: fix legacy order not displaying skill stats (#2283)

* chore: [auto-generated] Update urls.json [ci skip] (#2284)

Co-authored-by: magicus <magicus@users.noreply.github.com>

* feat: Pet menu search (#2285)

* feat: Pet menu search

* ci: spotless formatting

---------

Co-authored-by: DonkeyBlaster <DonkeyBlaster@users.noreply.github.com>

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: magicus <magicus@users.noreply.github.com>
Co-authored-by: Ryan <57310593+DonkeyBlaster@users.noreply.github.com>
Co-authored-by: DonkeyBlaster <DonkeyBlaster@users.noreply.github.com>

ryanzhoudev committed:
loadout list test

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
minor cleanup

ryanzhoudev committed:
summary box

ryanzhoudev committed:
selected ui

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
cool new save format, loading also works now

ryanzhoudev committed:
some stuff for full builds

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
fix delete/load buttons visibility, spotless, dont reload page on save

ryanzhoudev committed:
go by 5's

ryanzhoudev committed:
cleaner loop repetition

ryanzhoudev committed:
don't refresh on delete, parse entire sp after load

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
fix no-tome problems

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
colored widget text for level, some weird thenIf thing I came up with

ryanzhoudev committed:
that was terrible, go back to global var

ryanzhoudev committed:
Merge branch 'main' into sploadout

ryanzhoudev committed:
fix non-builds not saving, confirm to overwrite, layout fixes

ryanzhoudev committed:
clean up duplication

ryanzhoudev committed:
docs, spotless

ryanzhoudev committed:
fix weird inverted isBuild bug, layout impovements, both way converts

ryanzhoudev committed:
fix double confirm message bug

kristofbolyai commented on common/src/main/java/com/wynntils/utils/wynn/ContainerUtils.java:
```diff
@@ -48,7 +48,8 @@ public static boolean openInventory(int slotNum) {
      */
-    public static void clickOnSlot(int clickedSlot, int containerId, int mouseButton, List<ItemStack> items) {
+    public static void clickOnSlot(
+            int clickedSlot, int containerId, int mouseButton, List<ItemStack> items, boolean shiftClick) {
```
I would create a method with shift-click being off by default, to reduce the changes in the PR, and to make the method simpler to use.

kristofbolyai commented on common/src/main/java/com/wynntils/utils/mc/McUtils.java:
```diff
@@ -104,4 +105,12 @@ public static void sendSequencedPacket(PredictiveAction predictiveAction) {
+    public static void closeBackgroundContainer() {
+        player().connection.send(new ServerboundContainerClosePacket(McUtils.player().containerMenu.containerId));
+        player().containerMenu = McUtils.player().inventoryMenu;
+    }
```
Maybe fits into container utils more, this is not so commonly used to be in McUtils

kristofbolyai commented on common/src/main/java/com/wynntils/features/utilities/SkillPointLoadoutsFeature.java:
```diff
@@ -0,0 +1,46 @@
+import net.minecraft.network.chat.Component;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class SkillPointLoadoutsFeature extends Feature {
```
Missing ConfigCategory

kristofbolyai commented on common/src/main/java/com/wynntils/features/utilities/SkillPointLoadoutsFeature.java:
```diff
@@ -0,0 +1,46 @@
+    public void onCharacterInfoScreenOpened(ScreenOpenedEvent e) {
+        if (!(e.getScreen() instanceof ContainerScreen screen)) return;
+        if (!Models.Container.isCharacterInfoScreen(screen)) return;
+        if (screen.topPos == 0 || screen.width == 0) return; // Not done loading
```
ScreenOpenedEvent has .Pre and .Post, perhaps you need to use post?

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SavableSkillPointSet.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.Collections;
+import java.util.List;
+
+public class SavableSkillPointSet {
```
This class really wants to be a record, doesn't it? You can create factory methods instead of the constructors.

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+
+public class SkillPointModel extends Model {
+    @Persisted
+    private final Storage<Map<String, SavableSkillPointSet>> skillPointLoadouts = new Storage<>(new LinkedHashMap<>());
```
The persisted wrappers default to `TreeMap`, and you can't change that at the moment. I assume you wanted  to keep the order here, so `TreeMap` should work just fine.

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+    /**
+     * Saves the current equipped gear and provided skill points.
+     * @param name
+     * @param skillPoints
```
missing descriptions

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+        List<String> accessoryNames = new ArrayList<>();
+        for (int i : ACCESSORY_SLOTS) {
+            ItemStack itemStack = McUtils.inventory().getItem(i);
+            if (!itemStack.isEmpty() && !itemStack.getHoverName().getString().equals("§7Accessory Slot")) {
```
Might as well extract this to a const field: `§7Accessory Slot`

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SavableSkillPointSet.java:
```diff
@@ -0,0 +1,82 @@
+        this.defence = skillPoints[3];
+        this.agility = skillPoints[4];
+        this.armourNames = new ArrayList<>(armourNames);
+        this.accessoryNames = new ArrayList<>(accessoryNames);
```
These should be immutable lists to prevent any weird cases. That also means you can cache `isBuild`. Whenever you convert a saved skill set to a build, just recreate this object.

kristofbolyai commented on common/src/main/java/com/wynntils/handlers/container/scriptedquery/QueryBuilder.java:
```diff
@@ -43,6 +43,13 @@ public QueryBuilder then(QueryStep step) {
+            steps.add(step);
+        }
+        return this;
+    }
```
I'll get back to this, as we've discussed.

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+
+        ScriptedContainerQuery query = ScriptedContainerQuery.builder("Loading Skill Point Loadout Query")
+                .onError(msg -> WynntilsMod.warn("Failed to load skill point loadout: " + msg))
+                .then(QueryStep.useItemInHotbar(CharacterModel.CHARACTER_INFO_SLOT - 1)
```
I don't think we need the -1 anywhere else?

kristofbolyai replied:
Oh, I see now! I think you confused this variable. This variable is used for getting the player info item from the character menu, not to open the character info menu. So just introduce a new field with the correct index.
![image](https://github.com/Wynntils/Artemis/assets/49001742/d1c56b59-db64-467a-8eea-03ecfc2ef321)

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+        // we need to figure out which points we can subtract from first to actually allow assigning for positive points
+        Map<Skill, Integer> negatives = new EnumMap<>(Skill.class);
+        Map<Skill, Integer> positives = new EnumMap<>(Skill.class);
+        for (int i = 0; i < 5; i++) {
```
```suggestion
        for (int i = 0; i < Skill.values().length; i++) {
```

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+            }
+        }
+
+        AtomicBoolean confirmationCompleted = new AtomicBoolean(false);
```
If  you used regular loops, you would not need atomics.

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+
+    public void deleteLoadout(String name) {
+        skillPointLoadouts.get().remove(name);
+    }
```
can you put after the save method?

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+     * Closes any open containers (but not the screen shown) then queries compass (and tome menus) depending on if
+     * tomes have been unlocked.
+     */
+    public void populateSkillPoints() {
```
can you put this to the public methods, like after loadLoadout?

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+        McUtils.closeBackgroundContainer();
+
+        Managers.TickScheduler.scheduleNextTick(() -> {
+            assignedSkillPoints.clear();
```
I would rather you just set it to a new instance, which is usually safer in case something holds a reference to the object, we won't crash with CCME.

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+
+    private void calculateGearSkillPoints() {
+        gearSkillPoints.clear();
+        craftedSkillPoints.clear();
```
Same here

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+                    && !itemStack.getHoverName().getString().equals("§7Accessory Slot")) {
+                WynntilsMod.warn("Skill Point Model failed to parse accessory: "
+                        + LoreUtils.getStringLore(McUtils.inventory().getItem(i)));
+            }
```
You can extract the highlighted part in both loops into a single method (taking the ItemStack as the arg)

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+     */
+    private void queryTotalAndTomeSkillPoints() {
+        totalSkillPoints.clear();
+        tomeSkillPoints.clear();
```
Same here, new instance should be better

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+    }
+
+    private void processTomeSkillPoints(ContainerContent content) {
+        tomeSkillPoints.clear();
```
.clear

kristofbolyai commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,386 @@
+
+    public Map<String, SavableSkillPointSet> getLoadouts() {
+        return skillPointLoadouts.get();
+    }
```
public methods should be before private ones

kristofbolyai commented on common/src/main/java/com/wynntils/screens/skillpointloadouts/SkillPointLoadoutsScreen.java:
```diff
@@ -0,0 +1,577 @@
+                            // in case user pressed on both buttons
+                            saveBuildButton.setMessage(
+                                    Component.translatable("screens.wynntils.skillPointLoadouts.saveBuild"));
+                        }
```
We've always just created new classes extending WynntilsButton, instead of using anonymous classes. It is a bit of boilerplate, but it improves this class' readability quite a bit.

kristofbolyai commented on common/src/main/java/com/wynntils/screens/skillpointloadouts/SkillPointLoadoutsScreen.java:
```diff
@@ -0,0 +1,577 @@
+    }
+
+    private void populateLoadouts() {
+        loadoutWidgets.clear();
```
better to set a new instance here, as well

ryanzhoudev committed:
new method for shift clicks, move closeBackgroundContainer, .post

ryanzhoudev committed:
record, treemap, javadoc, accessory const

ryanzhoudev committed:
cache isbuild, spotless

ryanzhoudev committed:
regular loops, move methods around

ryanzhoudev committed:
remove .clear

ryanzhoudev committed:
fix fast queries not waiting for container completion

ryanzhoudev committed:
fix confirm button not disappearing

ryanzhoudev committed:
move publics up

ryanzhoudev committed:
extract method for gear

ryanzhoudev committed:
named button class, spotless

ryanzhoudev committed:
Merge branch 'main' into sploadout

magicus commented on common/src/main/java/com/wynntils/models/skillpoint/SavableSkillPointSet.java:
```diff
@@ -0,0 +1,78 @@
+        return (int) Math.ceil(getSkillPointsSum() / 2.0) + 1;
+    }
+
+    public boolean isBuild() {
```
This is true if we have any armor or accessories..? Why store it as a boolean?

kristofbolyai replied:
Technically you don't even need this method, since the boolean is already "cached"

magicus replied:
Sorry, my comments were a bit tarse, I am lying in bed in a strange position and had diffuculty typing. :)

What I meant was that it seemed to me that this value was set to false iff both armor and accessories where empty. But this can as easily be checked on the arrays, and it does not seem to be necessary to cache this in a boolean field; the isBuild() method can check the two arrays directly.

magicus commented on common/src/main/java/com/wynntils/utils/wynn/ContainerUtils.java:
```diff
@@ -66,7 +66,31 @@ public static void clickOnSlot(int clickedSlot, int containerId, int mouseButton
+    /**
+     * Closes invisible containers opened in the background, without closing the visible screen.
+     */
+    public static void closeBackgroundContainer() {
```
This sounds scary...

ryanzhoudev replied:
Can't use queries without it, unless we remove the open container check. It's still "opened" in the background when our custom screen is on top

magicus replied:
I'm too tired to fully understand this now. But if @kristofbolyai says it is fine, I guess I am too.

kristofbolyai replied:
I have a round of reviewing specifically for the query system changes to do, I'll make sure to check if this is the best we can do.

magicus commented on common/src/main/java/com/wynntils/models/skillpoint/SkillPointModel.java:
```diff
@@ -0,0 +1,388 @@
+ * Copyright © Wynntils 2023.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.models.skillpoint;
```
I'm thinking this belongs in `models.character`...

ryanzhoudev replied:
Honestly I think we already have too much irrelevant stuff in there... At first I was under the impression that CharacterModel meant what we would previously call ClassModel, just the data about the player's currently selected class. But then slowly we added death location, Silverbull, veteran stuff, etc.

kristofbolyai replied:
You keep the model separate, just move the package, as it's still character related. That way the code is not getting messier. At least, I think this is what magicus meant.

magicus replied:
Yes. While we often has the format "models.foo" contains "FooModel", I don't think this is really necessary. The package grouping is on a somewhat higher level, and can contain many independent, but related, models.

I thought the CharacterModel contained info about our current skill points, but maybe I remembered wrong? The skill point loadout should be located next to the model containing info on our current skill points anyway. If it is not the CharacterModel, then whatever model it is.

kristofbolyai replied:
> I thought the CharacterModel contained info about our current skill points, but maybe I remembered wrong?

There is only a FIXME about adding support for it :)

magicus replied:
*lol* 

Then maybe that fixme can be removed now? Or, at least if the skillpoint model is flexible enough to allow us to read the current values of our skill points. (I'm thinking they should basically be exported by the model so they can e.g. be available as functions).

ryanzhoudev committed:
remove extra method

ryanzhoudev committed:
move to character

magicus committed:
[ci skip] chore: Add framework for unit testing of regexes (#2303)

magicus committed:
[ci skip] chore: Add unit testing to GHA (#2306)

magicus committed:
[ci skip] chore: Move all regex testing from regexr.com to unit tests (#2305)

* chore: Move tests from regexr

* chore: Move more tests from regexr

* chore: Move more tests from regexr

* chore: Add support for enums

* chore: Always put titles before match line

* chore: Move more tests from regexr

* chore: Move more tests from regexr

* chore: Move the final tests from regexr

---------

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

kristofbolyai committed:
chore: Don't run tests on "build" (#2307)

Turns out we already ran tests, when we built the mod. I've made sure we only run them separately now. (2 second runtime for tests was a little suspicious)

ryanzhoudev committed:
conditionalThen for query steps

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

magicus commented on common/src/main/java/com/wynntils/commands/BombCommand.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright © Wynntils 2022-2023.
+ * Copyright © Wynntils 2022-2024.
```
How come all these files have been touched by spotless?

kristofbolyai committed:
Merge branch 'main' into sploadout

# Conflicts:
#	fabric/src/test/java/TestRegex.java
